### PR TITLE
Fix #693: Allow multiple application IDs for operations

### LINKED ISSUE #693: Allow multiple application IDs for operations
The operation can be application independent if it chooses so. Basically, we might want to push approve payment in mtoken or mobile banking, optionally filtered by the operation flag.

### PR DESCRIPTION
**Pull request by petrdvorak**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -114,7 +115,7 @@
 
         <!-- Test Dependencies -->
         <junit-jupiter-engine.version>5.8.2</junit-jupiter-engine.version>
-        <h2.version>2.1.210</h2.version>
+        <h2.version>1.4.200</h2.version>
```
Comment (banterCZ) [inline pom.xml:118]: Why downgraded?

File: powerauth-admin/pom.xml [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    <description>PowerAuth Server Admin Console</description>
+    <artifactId>powerauth-admin</artifactId>
+    <version>1.3.0-SNAPSHOT</version>
+    <packaging>war</packaging>
+
+    <parent>
+        <groupId>io.getlime.security</groupId>
+        <artifactId>powerauth-server-parent</artifactId>
+        <version>1.3.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
```
Comment (banterCZ) [inline powerauth-admin/pom.xml:16]: Isn't this implicit relative path?

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/validator/OperationApproveRequestValidator.java [hunk 3]
```diff
@@ -31,8 +32,8 @@ public static String validate(OperationApproveRequest source) {
         if (source == null) {
             return "Operation approve request must not be null";
         }
-        if (source.getApplicationId() == null) {
-            return "Application ID must not be null when creating operation";
+        if (source.getApplicationId() == null || !StringUtils.hasText(source.getApplicationId())) {
```
Comment (banterCZ) [inline powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/validator/OperationApproveRequestValidator.java:35]: `source.getApplicationId() == null` is already part of `StringUtils`, isn't it?

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/validator/OperationCreateRequestValidator.java [hunk 4]
```diff
@@ -31,8 +31,8 @@ public static String validate(OperationCreateRequest source) {
         if (source == null) {
             return "Operation create request must not be null";
         }
-        if (source.getApplicationId() == null) {
-            return "Application ID must not be null when creating operation";
+        if (source.getApplications() == null || source.getApplications().size() == 0) {
```
Comment (banterCZ) [inline powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/validator/OperationCreateRequestValidator.java:34]: Posible to use `CollectionUtils`.

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/validator/OperationApproveRequestValidator.java (hunk 3 above)
Comment (saalistaja) [inline powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/validator/OperationApproveRequestValidator.java:35]: `hasText` is null safe

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/model/entity/ApplicationEntity.java [hunk 5]
```diff
@@ -126,18 +132,38 @@ public List<ApplicationVersionEntity> getVersions() {
...
+
+    @Override
+    public String toString() {
+        return "ApplicationEntity{" +
+                "rid=" + rid +
+                ", id='" + id + '\'' +
+                ", roles=" + roles +
+                ", versions=" + versions +
+                ", callbacks=" + callbacks +
+                ", recoveryCodes=" + recoveryCodes +
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/model/entity/ApplicationEntity.java:166]: Is it safe to expose recoveryCodes at toString?

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java [hunk 6]
```diff
@@ -861,14 +858,14 @@ public LookupApplicationByAppKeyResponse lookupApplicationByAppKey(LookupApplica
     @Override
     @Transactional
     public CreateApplicationResponse createApplication(CreateApplicationRequest request) throws GenericServiceException {
-        if (request.getApplicationName() == null) {
+        if (request.getApplicationId() == null) {
             logger.warn("Invalid request parameter applicationName in method createApplication");
```
Comment (saalistaja) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java:862]: `applicationName` survived in the log message?

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java [hunk 7]
```diff
@@ -861,14 +858,14 @@ public LookupApplicationByAppKeyResponse lookupApplicationByAppKey(LookupApplica
...
             // Rollback is not required, error occurs before writing to database
             throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_REQUEST);
         }
         try {
-            logger.info("CreateApplicationRequest received, application name: {}", request.getApplicationName());
-            final CreateApplicationResponse response = behavior.getApplicationServiceBehavior().createApplication(request.getApplicationName(), keyConvertor);
+            logger.info("CreateApplicationRequest received, application name: {}", request.getApplicationId());
```
Comment (saalistaja) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java:867]: dtto

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java [hunk 8]
```diff
@@ -886,14 +883,14 @@ public CreateApplicationResponse createApplication(CreateApplicationRequest requ
     @Override
     @Transactional
     public CreateApplicationVersionResponse createApplicationVersion(CreateApplicationVersionRequest request) throws GenericServiceException {
-        if (request.getApplicationVersionName() == null) {
+        if (request.getApplicationId() == null) {
             logger.warn("Invalid request parameter applicationVersionName in method createApplicationVersion");
```
Comment (saalistaja) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java:887]: dtto

File: powerauth-rest-client-spring/src/main/java/com/wultra/security/powerauth/rest/client/PowerAuthRestClient.java [hunk 9]
```diff
@@ -232,22 +232,22 @@ public InitActivationResponse initActivation(InitActivationRequest request, Mult
     }
 
     @Override
-    public InitActivationResponse initActivation(String userId, Long applicationId) throws PowerAuthClientException {
+    public InitActivationResponse initActivation(String userId, String applicationId) throws PowerAuthClientException {
```
Comment (banterCZ) [inline powerauth-rest-client-spring/src/main/java/com/wultra/security/powerauth/rest/client/PowerAuthRestClient.java:235]: This is not backward compatible.

**Review by saalistaja — APPROVED** [review]
Looks good. I have advice only minor code changes.

File: pom.xml (hunk 1 above)
Comment (petrdvorak) [reply pom.xml:118]: The tests do not work with the new version - I think we upgraded too proactively.

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/validator/OperationApproveRequestValidator.java (hunk 3 above)
Comment (petrdvorak) [reply powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/validator/OperationApproveRequestValidator.java:35]: Good catch.

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/validator/OperationCreateRequestValidator.java (hunk 4 above)
Comment (petrdvorak) [reply powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/validator/OperationCreateRequestValidator.java:34]: Let's change this later.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java (hunk 6 above)
Comment (petrdvorak) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java:862]: Good catch

File: powerauth-rest-client-spring/src/main/java/com/wultra/security/powerauth/rest/client/PowerAuthRestClient.java (hunk 9 above)
Comment (petrdvorak) [reply powerauth-rest-client-spring/src/main/java/com/wultra/security/powerauth/rest/client/PowerAuthRestClient.java:235]: Yes, this is the trade off - we will document it and announce this on version release.

File: powerauth-admin/pom.xml (hunk 2 above)
Comment (petrdvorak) [reply powerauth-admin/pom.xml:16]: Not sure - we specify this in every child pom.xml file - we can review this separately.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java [hunk 10]
```diff
@@ -111,13 +111,12 @@ public OperationDetailResponse createOperation(OperationCreateRequest request) t
...
         final OperationTemplateEntity templateEntity = template.get();
 
-        // Check if application exists
-        final Optional<ApplicationEntity> application = applicationRepository.findById(applicationId);
-        if (!application.isPresent()) {
-            logger.error("Application was not found for ID: {}", applicationId);
+        // Check if applications exist
+        final List<ApplicationEntity> applicationEntities = applicationRepository.findAllByIdIn(applications);
+        if (applicationEntities.size() != applications.size()) {
+            logger.error("Application was not found for ID: {}", applications);
```
[the remark this answers is not in the document]
Comment (petrdvorak) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java:117]: Improved

**Pull request merged**